WSFrontend/org.openl.rules.ruleservice.ws.storelogdata.db/src/org/openl/rules/ruleservice/storelogdata/db/EntityManagerOperations.java

```diff
@@ -59,16 +54,28 @@ public int hashCode() {
     private final AtomicReference<Map<Key, EntityManagerFactory>> entityManagers = new AtomicReference<>(
         Collections.unmodifiableMap(new HashMap<>()));
 
-    public void save(Class<?>[] entityClasses, Object entity) throws PersistenceException {
+    public void save(Class<?>[] entityClasses, Object entity) {
         if (entity == null) {
             return;
         }
-        EntityManager entityManager = getEntityManagerFactory(entityClasses).createEntityManager();
-        EntityTransaction tx = entityManager.getTransaction();
-        tx.begin();
-        entityManager.merge(entity);
-        tx.commit();
-        entityManager.close();
+        EntityManager entityManager = null;
+        EntityTransaction tx = null;
+        try {
+            entityManager = getEntityManagerFactory(entityClasses).createEntityManager();
+            tx = entityManager.getTransaction();
+            tx.begin();
+            entityManager.merge(entity);
+            tx.commit();
+        } catch (RuntimeException e) {
+            if (tx != null && tx.isActive()) {
+                tx.rollback();
+            }
+            throw e;
+        } finally {
+            if (entityManager != null) {
+                entityManager.close();
+            }
+        }
     }
 
     public EntityManagerFactory getEntityManagerFactory(Class<?>[] entityClasses) {
```

Comment:
I thing `EntityManagerFactory` can be replaced with `org.hibernate.SessionFactory` and use `openSession()` instead of `createEntityManager()`. It will allow to use try-with-resource statement.

Reply:
done